broadway_api/callbacks/worker.py

```diff
@@ -15,16 +18,88 @@ def worker_heartbeat_callback(settings):
     """
     heartbeat_timestamp = get_time()
     heartbeat_interval = settings["CONFIG"]["HEARTBEAT_INTERVAL"]
+    conn_map = settings["WS_CONN_MAP"]
 
     dao = WorkerNodeDao(settings)
+
     for node in dao.find_by_liveness(alive=True):
         if (
             heartbeat_timestamp - node.last_seen
         ).total_seconds() >= 2 * heartbeat_interval:
+            if node.use_ws and node.id in conn_map:
+                conn_map[node.id].close()
+
             _handle_lost_worker_node(settings, node)
 
+        elif node.use_ws and node.id in conn_map:
+            # send ping messages to all connections
+            # excepting pong response before the next checking point
+            conn_map[node.id].ping()
+
+
+def worker_lost_callback(settings, worker_id, reason="closed connection"):
+    dao = WorkerNodeDao(settings)
+    worker = dao.find_by_id(worker_id)
+
+    if worker is None:
+        logger.critical("dead worker {} not found".format(worker_id))
+        return
+
+    _handle_lost_worker_node(settings, worker, reason=reason)
```

Comment:
Do we also want to call `conn_map[node.id].close()` here since the worker can be lost without closing connection?

Reply:
I think this function is only used by the `on_close` method of `WorkerConnectionHandler`, which should only be called when a connection is being closed.